torch/utils/checkpoint.py

```diff
@@ -0,0 +1,114 @@
+import torch
+
+
+def detach_variable(inputs):
+    if torch.is_tensor(inputs):
+        inp = inputs.detach()
+        return inp
+    elif isinstance(inputs, tuple):
+        return tuple(detach_variable(v) for v in inputs)
+    else:
+        raise RuntimeError(
+            "Only tensor or tuple of tensors is supported. Got Unsupported input type: ", type(inputs).__name__)
+
+
+class CheckpointFunction(torch.autograd.Function):
+
+    @staticmethod
+    def forward(ctx, run_function, *args):
+        ctx.run_function = run_function
+        ctx.save_for_backward(*args)
+        with torch.no_grad():
+            outputs = run_function(*args)
+        return outputs
+
+    @staticmethod
+    def backward(ctx, *args):
+        if (not torch.autograd.is_checkpoint_valid()):
+            raise RuntimeError("Checkpointing is not compatible with .grad(), please use .backward() if possible")
+        inputs = ctx.saved_tensors
+        inputs_list = detach_variable(inputs)
+        with torch.enable_grad():
+            outputs = ctx.run_function(*inputs_list)
+
+        if torch.is_tensor(outputs):
```

Comment:
Please use `isinstance`

Reply:
can you comment more on why using `torch.is_tensor` incorrect? are there any cases that this wouldn't handle or are we deprecating it? happy to make the change but want to understand a bit better :)

Reply:
`isinstance` is just more idiomatic in Python code